tokio/src/util/slab.rs

```diff
@@ -0,0 +1,745 @@
+                slots: Vec::new(),
+                head: 0,
+                used: 0,
+            }),
+            len: 0,
+            prev_len: 0,
+        }
+    }
+}
+
+impl<T> Page<T> {
+    /// Release a slot into the page's free list
+    fn release(&self, value: *const Value<T>) {
+        let mut locked = self.slots.lock().unwrap();
+
+        let idx = locked.index_for(value);
+        locked.slots[idx].next = locked.head as u32;
+        locked.head = idx;
+        locked.used -= 1;
+
+        self.used.store(locked.used, Release);
+    }
+}
+
+impl<T> CachedPage<T> {
+    /// Refresh the cache
+    fn refresh(&mut self, page: &Page<T>) {
+        let slots = page.slots.lock().unwrap();
+        self.slots = slots.slots.as_ptr();
+        self.init = slots.slots.len();
+    }
+
+    // Get a value by index
+    fn get(&self, idx: usize) -> &T {
+        assert!(idx < self.init);
+
+        // Safety: Pages are allocated concurrently, but are only ever
+        // **deallocated** by `Slab`. `Slab` will always have a more
+        // conservative view on the state of the slot array. Once `CachedPage`
+        // sees a slot pointer and initialized offset, it will remain valid
+        // until `compact()` is called. The `compact()` function also updates
+        // `CachedPage`.
+        unsafe {
+            let slot = self.slots.offset(idx as isize);
+            let value = slot as *const Value<T>;
+
+            &(*value).value
+        }
+    }
+}
+
+impl<T> Default for CachedPage<T> {
+    fn default() -> CachedPage<T> {
+        CachedPage {
+            slots: ptr::null(),
+            init: 0,
+        }
+    }
+}
+
+impl<T> Slots<T> {
+    /// Maps a slot pointer to an offset within the current page.
+    ///
+    /// # Panics
+    ///
+    /// panics if the provided slot pointer is not contained by the page.
+    fn index_for(&self, slot: *const Value<T>) -> usize {
+        use std::mem;
+
+        let base = &self.slots[0] as *const _ as usize;
+
+        assert!(base != 0, "page is unallocated");
+
+        let slot = slot as usize;
+        let width = mem::size_of::<Slot<T>>();
+
+        assert!(slot >= base, "unexpected pointer");
+
+        let idx = (slot - base) / width;
+        assert!(idx < self.slots.len() as usize);
+
+        idx
+    }
+}
+
+impl<T: Entry> Slot<T> {
+    // Generates a `Ref` for the slot. This involves bumping the page's ref count.
+    fn gen_ref(&self, page: &Arc<Page<T>>) -> Ref<T> {
+        // The ref holds a ref on the page.
+        mem::forget(page.clone());
+        let slot = self as *const Slot<T>;
+        let value = slot as *const Value<T>;
+
+        Ref { value }
+    }
+}
+
+impl<T> Value<T> {
+    // Release the slot, returning the `Arc<Page<T>>` logically owned by the ref.
+    fn release(&self) -> Arc<Page<T>> {
+        // Safety: called by `Ref`, which owns an `Arc<Page<T>>` instance.
+        let page = unsafe { Arc::from_raw(self.page) };
```

Comment:
It would be nice to have a comment somewhere that summarizes the `Arc` dance that's going on here.

Reply:
I expanded the comment.